File: src/domain/stateMachine/DrivingStateProcessor.ts
import type { DrivingState, TelemetryPoint } from '../models/telemetry';

export interface StateMachineConfig {
  drivingSpeedThresholdKmH: number; // 5.0 km/h
  idlingSpeedThresholdKmH: number;  // 3.0 km/h
  idlingTimeThresholdSec: number;   // 15 seconds
  pitStopTimeThresholdSec: number;  // 120 seconds (2 mins)
}

export const DEFAULT_STATE_CONFIG: StateMachineConfig = {
  drivingSpeedThresholdKmH: 5.0,
  idlingSpeedThresholdKmH: 3.0,
  idlingTimeThresholdSec: 15,
  pitStopTimeThresholdSec: 120,
};

export class DrivingStateProcessor {
  private config: StateMachineConfig;
  private currentState: DrivingState = 'OFFLINE';
  private stationaryStartTime: number | null = null;

  constructor(config: Partial<StateMachineConfig> = {}) {
    this.config = { ...DEFAULT_STATE_CONFIG, ...config };
  }

  public getCurrentState(): DrivingState {
    return this.currentState;
  }

  public processTelemetry(point: TelemetryPoint): DrivingState {
    const speed = point.speedKmH;
    const now = point.timestamp;

    if (speed >= this.config.drivingSpeedThresholdKmH) {
      // Vehicle is moving fast enough -> DRIVING state
      this.stationaryStartTime = null;
      this.currentState = 'DRIVING';
      return this.currentState;
    }

    if (speed < this.config.idlingSpeedThresholdKmH) {
      // Vehicle speed is low
      if (this.stationaryStartTime === null) {
        this.stationaryStartTime = now;
      }

      const stationaryDurationSec = (now - this.stationaryStartTime) / 1000;

      if (stationaryDurationSec >= this.config.pitStopTimeThresholdSec) {
        // Stationary for > 2 mins -> PIT_STOP
        this.currentState = 'PIT_STOP';
      } else if (stationaryDurationSec >= this.config.idlingTimeThresholdSec) {
        // Stationary for > 15s -> IDLING
        this.currentState = 'IDLING';
      } else {
        // Transitional state, keep previous state unless offline
        if (this.currentState === 'OFFLINE') {
          this.currentState = 'IDLING';
        }
      }
      return this.currentState;
    }

    // Intermediate speed (between 3km/h and 5km/h)
    return this.currentState === 'OFFLINE' ? 'DRIVING' : this.currentState;
  }

  public reset(): void {
    this.currentState = 'OFFLINE';
    this.stationaryStartTime = null;
  }
}

File: smart_trip_ai_mobile/lib/domain/state_machine/driving_state_processor.dart
import '../models/telemetry.dart';

class StateMachineConfig {
  final double drivingSpeedThresholdKmH;
  final double idlingSpeedThresholdKmH;
  final double idlingTimeThresholdSec;
  final double pitStopTimeThresholdSec;

  const StateMachineConfig({
    this.drivingSpeedThresholdKmH = 5.0,
    this.idlingSpeedThresholdKmH = 3.0,
    this.idlingTimeThresholdSec = 15.0,
    this.pitStopTimeThresholdSec = 120.0,
  });
}

class DrivingStateProcessor {
  final StateMachineConfig config;
  DrivingState currentState = DrivingState.offline;
  int? stationaryStartTime;

  DrivingStateProcessor({this.config = const StateMachineConfig()});

  DrivingState getCurrentState() {
    return currentState;
  }

  DrivingState processTelemetry(TelemetryPoint point) {
    final speed = point.speedKmH;
    final now = point.timestamp;

    if (speed >= config.drivingSpeedThresholdKmH) {
      // Vehicle is moving fast enough -> DRIVING state
      stationaryStartTime = null;
      currentState = DrivingState.driving;
      return currentState;
    }

    if (speed < config.idlingSpeedThresholdKmH) {
      // Vehicle speed is low
      stationaryStartTime ??= now;

      final stationaryDurationSec = (now - stationaryStartTime!) / 1000;

      if (stationaryDurationSec >= config.pitStopTimeThresholdSec) {
        // Stationary for > 2 mins -> PIT_STOP
        currentState = DrivingState.pitStop;
      } else if (stationaryDurationSec >= config.idlingTimeThresholdSec) {
        // Stationary for > 15s -> IDLING
        currentState = DrivingState.idling;
      } else {
        // Transitional state, keep previous state unless offline
        if (currentState == DrivingState.offline) {
          currentState = DrivingState.idling;
        }
      }
      return currentState;
    }

    // Intermediate speed (between 3km/h and 5km/h)
    return currentState == DrivingState.offline ? DrivingState.driving : currentState;
  }

  void reset() {
    currentState = DrivingState.offline;
    stationaryStartTime = null;
  }
}
